### PR TITLE
handling send to socket error

### LINKED ISSUE #27: SocketIO: Lots of panics on unwrap in case of there are lots of clients
error:
```
panicked at '14called `Result::unwrap()` on an `Err` value: PoisonError { .. }
', thread 'socketioxide/src/ns.rstokio-runtime-worker:' panicked at '104called `Result::unwrap()` on an `Err` value: PoisonError { .. }:', 14socketioxide/src/ns.rs
:104:14
```
How to reproduce:
run socketio server(i used main) `cargo run --release --bin socketioxide-e2e`
run js clients [client.zip](https://github.com/Totodore/socketioxide/files/11797832/client.zip)

For me it reproduces even with 100 clients.

also I locally tried to replace RWlock<Hashmap> to Dashmap and errors dissappear

### PR DESCRIPTION
**Pull request by biryukovmaxim**
fix #27 
create SendError enum, BroadcastError
handle these errors

@Totodore I left todos in places where I don't know how to handle error correctly

### CODE REVIEW

File: socketioxide/src/adapter.rs [hunk 1]
```diff
@@ -173,9 +173,14 @@ impl Adapter for LocalAdapter {
         let sockets = self.apply_opts(opts);
 
         tracing::debug!("broadcasting packet to {} sockets", sockets.len());
-        sockets
-            .into_iter()
-            .try_for_each(|socket| socket.send(packet.clone()))
+        sockets.into_iter().try_for_each(|socket| {
+            if let Err(SendError::SocketFull { .. }) = socket.send(packet.clone()) {
+                // todo skip message? return err? try send later? create delayed queue?
```
Comment (Totodore) [inline socketioxide/src/adapter.rs:178]: I would say, skip message and return a vec of errors ?
Or create a BroadcastError that can take a list of sids and the rejected packet.

File: socketioxide/src/errors.rs [hunk 2]
```diff
@@ -42,4 +48,33 @@ pub enum AckError {
...
+    SocketClosed { sid: Sid, packet: T },
+    #[error("error serializing json packet: {0:?}")]
+    Serialize(#[from] serde_json::Error),
+}
+
+impl<T: Debug> From<(TrySendError<T>, Sid)> for SendError<T> {
+    fn from((err, sid): (TrySendError<T>, Sid)) -> Self {
+        match err {
+            TrySendError::Closed(packet) => {
+                warn!("try to send to closed socket, sid: {sid}, packet: {packet:?}");
```
Comment (Totodore) [inline socketioxide/src/errors.rs:71]: Warn would log even when the DEBUG flag is off. We should switch to DEBUG

**Comment by Totodore** [comment]
I didn't realized that returning the packet to send when there is an error implies to `deserialize` it again or returning a raw `SendPacket`. These two solutions are not optimal...

We could tweak the emit fn so it could directly emit `SendPacket`s. Like that the user would be able to try to send the failed packet again

Otherwise, the easiest solution would be to return an error but without the packet...

What do you think @biryukovmaxim ?

**Review by Totodore — APPROVED** [review]
Merging this solution.
I don't see what we could do more to have better error handling with packet returned

**Comment by biryukovmaxim** [comment]
Noo plz, let me 20 minutes. I will show

**Comment by Totodore** [comment]
Sorry, I'm reopening the pr 😅

**Pull request merged**